diff -urp file-5.33.orig/magic/Magdir/python file-5.33/magic/Magdir/python
--- file-5.33.orig/magic/Magdir/python	2020-12-17 13:19:08.610803723 -0500
+++ file-5.33/magic/Magdir/python	2020-12-17 13:26:07.346954161 -0500
@@ -43,6 +43,18 @@
 !:strength + 15
 !:mime text/x-python
 
+0       search/1/wt     #!\ /usr/libexec/platform-python        Python script text executable
+!:strength + 15
+!:mime text/x-python
+
+0       search/1/wt     #!\ /usr/bin/python2    Python script text executable
+!:strength + 15
+!:mime text/x-python
+
+0       search/1/wt     #!\ /usr/bin/python3    Python script text executable
+!:strength + 15
+!:mime text/x-python
+
 
 # from module.submodule import func1, func2
 0	regex		\^from[\040\t\f\r\n]+([A-Za-z0-9_]|\\.)+[\040\t\f\r\n]+import.*$	Python script text executable
