diff --git a/magic/Magdir/archive b/magic/Magdir/archive
index 99798b0..945f536 100644
--- a/magic/Magdir/archive
+++ b/magic/Magdir/archive
@@ -150,7 +150,7 @@
 # Incremental snapshot gnu-tar format from:
 # https://www.gnu.org/software/tar/manual/html_node/Snapshot-Files.html
 0	string		GNU\ tar-	GNU tar incremental snapshot data
->&0	regex		[0-9]\.[0-9]+-[0-9]+	version %s
+>&0	regex		[0-9]\\.[0-9]+-[0-9]+	version %s
 
 # cpio archives
 #
diff --git a/magic/Magdir/ctf b/magic/Magdir/ctf
index ebea8f3..d91684d 100644
--- a/magic/Magdir/ctf
+++ b/magic/Magdir/ctf
@@ -20,4 +20,4 @@
 # CTF metadata (plain text)
 0	string	/*\x20CTF\x20   Common Trace Format (CTF) plain text metadata
 !:strength + 5			# this is to make sure we beat C
->&0	regex	[0-9]+\.[0-9]+	\b, v%s
+>&0	regex	[0-9]+\\.[0-9]+	\b, v%s
diff --git a/magic/Magdir/geo b/magic/Magdir/geo
index d72e514..dda5f73 100644
--- a/magic/Magdir/geo
+++ b/magic/Magdir/geo
@@ -1,6 +1,6 @@
 
 #------------------------------------------------------------------------------
-# $File: geo,v 1.7 2019/04/19 00:42:27 christos Exp $
+# $File: geo,v 1.8 2022/03/24 15:48:58 christos Exp $
 # Geo- files from Kurt Schwehr <schwehr@ccom.unh.edu>
 
 ######################################################################
@@ -28,8 +28,8 @@
 # Knudsen subbottom chirp profiler - Binary File Format: B9
 # KEB D409-03167 V1.75 Huffman
 0	string	KEB\ 	Knudsen seismic KEL binary (KEB) -
->4	regex	[-A-Z0-9]*	Software: %s
->>&1	regex	V[0-9]*\.[0-9]*	version %s
+>4	regex	[-A-Z0-9]+	Software: %s
+>>&1	regex	V[0-9]+\\.[0-9]+	version %s
 
 ######################################################################
 #
@@ -40,7 +40,7 @@
 
 # Caris LIDAR format for LADS comes as two parts... ascii location file and binary waveform data
 0	string	HCA	LADS Caris Ascii Format (CAF) bathymetric lidar
->4	regex [0-9]*\.[0-9]*	version %s
+>4	regex [0-9]+\\.[0-9]+	version %s
 
 0	string	HCB	LADS Caris Binary Format (CBF) bathymetric lidar waveform data
 >3      byte    x	version %d .
@@ -69,7 +69,7 @@
 
 # mb121 https://www.saic.com/maritime/gsf/
 8	string	GSF-v	SAIC generic sensor format (GSF) sonar data,
->&0	regex [0-9]*\.[0-9]*	version %s
+>&0	regex [0-9]+\\.[0-9]+	version %s
 
 # MGD77 - https://www.ngdc.noaa.gov/mgg/dat/geodas/docs/mgd77.htm
 # mb161
diff --git a/magic/Magdir/windows b/magic/Magdir/windows
index 8a7923f..1247e03 100644
--- a/magic/Magdir/windows
+++ b/magic/Magdir/windows
@@ -358,7 +358,7 @@
 # skip space at beginning
 >0     string		\040
 # name without extension and greater character or name with hlp extension
->>1	regex/c		\^([^\xd>]*|.*\.hlp)	MS Windows help file Content, based "%s"
+>>1	regex/c		\^([^\xd>]*|.*\\.hlp)	MS Windows help file Content, based "%s"
 !:mime	text/plain
 !:apple	????TEXT
 !:ext	cnt
@@ -535,7 +535,7 @@
 # http://www.winfaq.de/faq_html/Content/tip2500/onlinefaq.php?h=tip2653.htm
 # https://msdn.microsoft.com/en-us/library/windows/desktop/cc144102.aspx
 # .ShellClassInfo DeleteOnCopy LocalizedFileNames ASCII coded case-independent
->>&0	regex/c	\^(\.ShellClassInfo|DeleteOnCopy|LocalizedFileNames)]	Windows desktop.ini
+>>&0	regex/1024c	\^(\\.ShellClassInfo|DeleteOnCopy|LocalizedFileNames)]	Windows desktop.ini
 !:mime application/x-wine-extension-ini
 #!:mime text/plain
 # https://support.microsoft.com/kb/84709/
